File: frontend_compose/src/jsMain/kotlin/core/DI.kt
package core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Простейший DI для глобальных зависимостей.
 * Поддерживает стек и полиморфизм.
 */
object DI {
    val instances = mutableListOf<Any>()

    fun put(instance: Any) = instances.add(0, instance)

    fun putAll(instances: List<Any>) = instances.forEach { put(it) }

    suspend fun putAndInit(instance: Initializable) {
        put(instance)
        instance.init()
    }

    suspend fun putAndInitAll(instances: List<Initializable>) {
        withContext(Dispatchers.Main) { // избегаем многопоточности при инициализации зависимостей
            coroutineScope {
                instances.forEach { launch { putAndInit(it) } }
            }
        }
    }

    inline fun <reified T> get(): T {
        for (i in instances) {
            if (i is T) return i
        }
        throw Exception("Instance of ${T::class.simpleName} is not found in global DI")
    }
}

/**
 * Краткий синоним для DI.get()
 */
inline fun <reified T> di(): T = DI.get()

/**
 * Асинхронный инициализатор для глобальных объектов.
 * Должен запускаться при старте главной страницы, с выводом ошибок туда же.
 */
interface Initializable {
    suspend fun init()
}

/**
 * Пример использования с полиморфизмом
 */
private abstract class AbstractUserSession
private class MyUserSession : AbstractUserSession()

fun exampleOfUseDI() {
    DI.put(MyUserSession())
    println(di<AbstractUserSession>())
    println(di<MyUserSession>())
}

File: frontend_flutter/lib/core/di.dart
/// Простейший DI для глобальных зависимостей.
/// Используется в слоях, где недоступен BuildContext.
/// Поддерживает стек и полиморфизм.
class DI {
  static final List<Object> _instances = [];

  static void put(Object instance) => _instances.insert(0, instance);

  static void putAll(List<Object> instances) => instances.forEach(put);

  static Future<void> putAndInit(Initializable instance) {
    _instances.insert(0, instance);
    return instance.init();
  }

  static Future<void> putAndInitAll(List<Initializable> instances) {
    return Future.wait(instances.map(putAndInit).toList());
  }

  static T get<T>() {
    for (final i in _instances) {
      if (i is T) return i as T;
    }
    throw 'Instance of $T is not found in global DI';
  }
}

/// Краткий синоним для DI.get()
T di<T>() => DI.get<T>();

/// Асинхронный нициализатор для глобальных объектов.
/// Должен запускаться при старте главной страницы, с выводом ошибок туда же.
abstract class Initializable {
  Future<void> init();
}

/// Для использования в моделях, которые не являются ChangeNotifier,
/// но содержат внутри контроллеры, например TextEditingController и др.
abstract class Disposable {
  void dispose();
}

///-------------------------------------------------
/// Пример использования с полиморфизмом
void exampleOfUseGlobal() {
  DI.put(_MyUserSession());
  // ignore: avoid_print
  print(di<_AbstractUserSession>());
  // ignore: avoid_print
  print(di<_MyUserSession>());
}

class _AbstractUserSession {}

class _MyUserSession extends _AbstractUserSession {}
